### PR TITLE
Factor out common code between mooseObject and Action

### PR DESCRIPTION
**Pull request by GiudGiud**
So much factored from mooseobject we could almost make an Action a MooseObject!

This could be an alternate design but:
- it still keeps the class a little thick if we dont create interfaces
- there's a lot of parameters that dont belong for Action so we would not inherit the validParams

Another one could be to to use templating of the interfaces based on the base class MooseObject / Action.
But we would not gain much, since it already works without, at the cost of added code complexity

### CODE REVIEW

**Comment by GiudGiud** [comment]
I ll fix Griffin and the conflict once this is cleared

File: framework/include/actions/Action.h [hunk 1]
```diff
@@ -280,17 +152,14 @@ class Action : public ConsoleStreamInterface,
...
-}
+  using MooseBase::_app;
+  using MooseBase::_name;
+  using MooseBase::_type;
+  using MooseBaseErrorInterface::_console;
+  using MooseBaseErrorInterface::mooseDeprecated;
+  using MooseBaseErrorInterface::mooseError;
+  using MooseBaseErrorInterface::mooseWarning;
+  using MooseBaseParameterInterface::_action_factory;
+  using MooseBaseParameterInterface::_factory;
```
Comment (lindsayad) [inline framework/include/actions/Action.h:164]: why these? Are you changing the access level?

File: framework/include/base/MooseBaseErrorInterface.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+[[noreturn]] void callMooseErrorRaw(std::string & msg, MooseApp * app);
+
+/**
+ * Interface that provides APIs to output errors/warnings/info messages
+ */
+class MooseBaseErrorInterface : public ConsoleStreamInterface
+{
+public:
+  MooseBaseErrorInterface(const MooseBase * const base)
+    : ConsoleStreamInterface(base->getMooseApp()), _app(base->getMooseApp()), _moose_base(base){};
```
Comment (lindsayad) [inline framework/include/base/MooseBaseErrorInterface.h:28]:
```suggestion
    : ConsoleStreamInterface(base->getMooseApp()), _app(base->getMooseApp()), _moose_base(base) {}
```

File: framework/include/base/MooseObject.h [hunk 3]
```diff
@@ -55,297 +48,28 @@ class MooseObject : public ConsoleStreamInterface,
...
-  std::string errorPrefix(const std::string & error_type) const;
+  using MooseBaseErrorInterface::_console;
+  using MooseBaseErrorInterface::mooseDeprecated;
+  using MooseBaseErrorInterface::mooseError;
+  using MooseBaseErrorInterface::mooseWarning;
+  using MooseBaseParameterInterface::getCheckedPointerParam;
+  using MooseBaseParameterInterface::getParam;
+  using MooseBaseParameterInterface::paramError;
+  using MooseBaseParameterInterface::paramInfo;
+  using MooseBaseParameterInterface::paramWarning;
```
Comment (lindsayad) [inline framework/include/base/MooseObject.h:64]: Same why question as for Action

File: framework/include/base/MooseObject.h [hunk 4]
```diff
@@ -55,297 +48,28 @@ class MooseObject : public ConsoleStreamInterface,
...
-        param_prefix += ":";
-      return object_prefix + param_prefix + "\n    " + MooseUtils::replaceAll(msg, "\n", "\n    ");
-    }
-    return object_prefix + param_prefix + " " + msg;
-  }
+  using MooseBase::_app;
+  using MooseBase::_name;
+  using MooseBase::_type;
+  using MooseBaseParameterInterface::_action_factory;
+  using MooseBaseParameterInterface::_factory;
```
Comment (lindsayad) [inline framework/include/base/MooseObject.h:74]: and here. There's no template parameter dependent base, so this shouldn't be necessary from that regard

File: framework/src/actions/AddVariableAction.C [hunk 5]
```diff
@@ -235,8 +235,8 @@ AddVariableAction::createInitialConditionAction()
 std::string
 AddVariableAction::determineType(const FEType & fe_type, unsigned int components, bool is_fv)
 {
-  mooseDeprecated("AddVariableAction::determineType() is deprecated. Use "
-                  "AddVariableAction::variableType() instead.");
+  ::mooseDeprecated("AddVariableAction::determineType() is deprecated. Use "
```
Comment (lindsayad) [inline framework/src/actions/AddVariableAction.C:238]: why is this necessary? This is a inherited class method right?

File: framework/src/actions/AddVariableAction.C [hunk 6]
```diff
@@ -250,7 +250,7 @@ AddVariableAction::variableType(const FEType & fe_type, const bool is_fv, const
   {
     if (fe_type.family == LAGRANGE_VEC || fe_type.family == NEDELEC_ONE ||
         fe_type.family == MONOMIAL_VEC)
-      mooseError("Vector finite element families do not currently have ArrayVariable support");
+      ::mooseError("Vector finite element families do not currently have ArrayVariable support");
```
Comment (lindsayad) [inline framework/src/actions/AddVariableAction.C:253]: same question

File: framework/src/base/MooseBaseParameterInterface.C [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+#include "InputParameterWarehouse.h"
+
+std::string
+paramErrorPrefix(const InputParameters & params, const std::string & param)
+{
+  return params.errorPrefix(param);
+}
+
+MooseBaseParameterInterface::MooseBaseParameterInterface(const InputParameters & parameters,
+                                                         const MooseBase * base)
```
Comment (lindsayad) [inline framework/src/base/MooseBaseParameterInterface.C:25]:
```suggestion
                                                         const MooseBase * const base)
```

File: framework/src/base/MooseBaseParameterInterface.C [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+void
+MooseBaseParameterInterface::connectControllableParams(const std::string & parameter,
+                                                       const std::string & object_type,
+                                                       const std::string & object_name,
+                                                       const std::string & object_parameter) const
+{
+  MooseObjectParameterName primary_name(uniqueName(), parameter);
+  auto base_type = _factory.getValidParams(object_type).get<std::string>("_moose_base");
```
Comment (lindsayad) [inline framework/src/base/MooseBaseParameterInterface.C:40]:
```suggestion
  const auto base_type = _factory.getValidParams(object_type).get<std::string>("_moose_base");
```

File: framework/src/base/MooseBaseParameterInterface.C [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  MooseObjectParameterName secondary_name(base_type, object_name, object_parameter);
+  _moose_base->getMooseApp().getInputParameterWarehouse().addControllableParameterConnection(
+      primary_name, secondary_name);
+
+  const std::vector<std::string> & tags = _pars.get<std::vector<std::string>>("control_tags");
```
Comment (lindsayad) [inline framework/src/base/MooseBaseParameterInterface.C:45]:
```suggestion
  const auto & tags = _pars.get<std::vector<std::string>>("control_tags");
```

**Review by lindsayad — COMMENTED** [review]
a great refactor!

File: framework/src/actions/AddVariableAction.C (hunk 5 above)
Comment (GiudGiud) [reply framework/src/actions/AddVariableAction.C:238]: we re in a static function so we cant use a non-static member function

File: framework/include/actions/Action.h (hunk 1 above)
Comment (GiudGiud) [reply framework/include/actions/Action.h:164]: It seems I am because I needed this. Well this explains that, I just need to get the access level right in the base classes instead

**Review by lindsayad — APPROVED** [review]
